### PR TITLE
Add Decimal variant hash function

### PR DESCRIPTION
**Pull request by karteekmurthys**
The decimal variant hash function is required for the evaluation of ConstantTypedExpr.

### CODE REVIEW

File: velox/type/Variant.h [hunk 1]
```diff
@@ -152,6 +152,19 @@ struct DecimalCapsule {
     }
     return lhsIntegral < rhsIntegral;
   }
+
+  size_t hash() const {
+    VELOX_CHECK(
+        unscaledValue.has_value(), "Unscaled value must be initialized");
```
Comment (majetideepak) [inline velox/type/Variant.h:158]: `VELOX_CHECK(hasValue(), "Unscaled value is not initialized");` is simpler.

File: velox/type/Variant.cpp [hunk 2]
```diff
@@ -555,7 +555,14 @@ uint64_t variant::hash() const {
       return folly::Hash{}(
           timestampValue.getSeconds(), timestampValue.getNanos());
     }
-
+    case TypeKind::SHORT_DECIMAL: {
```
Comment (majetideepak) [inline velox/type/Variant.cpp:558]: Should we also extend the `isNull` check above for decimal values?

File: velox/type/Variant.h [hunk 3]
```diff
@@ -152,6 +152,19 @@ struct DecimalCapsule {
...
+    auto hasher = folly::Hash{};
+    auto unscaledValHash = hasher(unscaledValue.value());
```
Comment (majetideepak) [inline velox/type/Variant.h:160]: `value()` for simplicity.

File: velox/type/tests/VariantTest.cpp [hunk 4]
```diff
@@ -119,6 +139,7 @@ TEST(VariantTest, shortDecimalNull) {
   EXPECT_EQ(null.toJson(), "null");
   EXPECT_EQ(*null.inferType(), *DECIMAL(10, 5));
   EXPECT_THROW(variant::null(TypeKind::SHORT_DECIMAL), VeloxException);
+  EXPECT_THROW(null.value<TypeKind::SHORT_DECIMAL>().hash(), VeloxException);
```
Comment (majetideepak) [inline velox/type/tests/VariantTest.cpp:142]: looks like other types do support null values for hash.

Comment (karteekmurthys) [reply velox/type/tests/VariantTest.cpp:142]: What do you suggest the expected behavior to be?

File: velox/type/Variant.cpp (hunk 2 above)
Comment (karteekmurthys) [reply velox/type/Variant.cpp:558]: Are you suggesting to combine the cases for short and long and check for NULL here?

File: velox/type/Variant.cpp [hunk 5]
```diff
@@ -555,7 +561,14 @@ uint64_t variant::hash() const {
...
+      auto shortDecimalCapsule = value<TypeKind::SHORT_DECIMAL>();
```
Comment (majetideepak) [inline velox/type/Variant.cpp:565]: `const auto& shortDecimalCapsule` here and below to avoid a copy.

File: velox/type/tests/VariantTest.cpp [hunk 6]
```diff
@@ -111,14 +111,46 @@ TEST(VariantTest, shortDecimal) {
   // 0.1234 < 1.234
   EXPECT_LT(
       u2.value<TypeKind::SHORT_DECIMAL>(), v.value<TypeKind::SHORT_DECIMAL>());
+
+  // Hash test.
+  auto vValue = v.value<TypeKind::SHORT_DECIMAL>();
+  auto vHash = v.hash();
+  auto u2Hash = u2.hash();
```
Comment (majetideepak) [inline velox/type/tests/VariantTest.cpp:118]: this variable is used only once. use `u2.hash()` directly below.

File: velox/type/tests/VariantTest.cpp [hunk 7]
```diff
@@ -111,14 +111,46 @@ TEST(VariantTest, shortDecimal) {
...
-  EXPECT_EQ(*null.inferType(), *DECIMAL(10, 5));
+  variant n1 = variant::shortDecimal(std::nullopt, DECIMAL(10, 5));
+  EXPECT_TRUE(n1.isNull());
+  EXPECT_EQ(n1.toJson(), "null");
+  EXPECT_EQ(*n1.inferType(), *DECIMAL(10, 5));
   EXPECT_THROW(variant::null(TypeKind::SHORT_DECIMAL), VeloxException);
+
+  // n1 and n2 differ only in scale.
+  variant n2 = variant::shortDecimal(std::nullopt, DECIMAL(10, 4));
+  EXPECT_NE(n1.hash(), n2.hash());
```
Comment (majetideepak) [inline velox/type/tests/VariantTest.cpp:145]: use a variable for n1.hash() as it is used many times?

File: velox/type/tests/VariantTest.cpp [hunk 8]
```diff
@@ -111,14 +111,46 @@ TEST(VariantTest, shortDecimal) {
...
+
+  // Hash test.
+  auto vValue = v.value<TypeKind::SHORT_DECIMAL>();
+  auto vHash = v.hash();
+  auto u2Hash = u2.hash();
+  // u2 and v differ in only precision.
+  EXPECT_NE(vHash, u2Hash);
+
+  // v2 and v differ only in scale.
+  variant v2 =
```
Comment (majetideepak) [inline velox/type/tests/VariantTest.cpp:123]: Let's use auto here and below.

**Review by majetideepak — APPROVED** [review]
Thanks @karteekmurthys

**Review by mbasmanova — COMMENTED** [review]
@karteekmurthys @majetideepak 

> The decimal variant hash function is required for the compilation of FieldAccessTypedExpr.

Would you clarify where hash function is used in FieldAccessTypedExpr?

**Comment by majetideepak** [comment]
@mbasmanova below is the bt
image

**Comment by majetideepak** [comment]
It looks the expression is `ConstantTypedExpr`. @karteekmurthys can you confirm?

**Comment by mbasmanova** [comment]
> It looks like the expression is ConstantTypedExpr

That makes sense as ConstantTypedExpr stores the constant value as Variant. Let's update the PR description to fix the expression type and clarify that hash function is required to avoid run time, not compile time check.

File: velox/type/Variant.h [hunk 9]
```diff
@@ -152,6 +152,17 @@ struct DecimalCapsule {
     }
     return lhsIntegral < rhsIntegral;
   }
+
+  size_t hash() const {
+    auto hasher = folly::Hash{};
+    auto unscaledValHash = hasValue() ? hasher(value()) : 0;
+    auto precisionHash = hasher(precision);
+    auto scaleHash = hasher(scale);
+    auto combinedPrecisionScaleHash =
```
Comment (mbasmanova) [inline velox/type/Variant.h:161]: nit: combinedPrecisionScaleHash name is very long; perhaps, use 'hash'?

```
auto hash = folly::hash::hash_combine_generic(hasher, hasher(precision), hasher(scale));
if (hasValue()) {
   hash = folly::hash::hash_combine_generic(hasher, hasher(value()), hash);
}
return hash;
```

File: velox/type/tests/VariantTest.cpp [hunk 10]
```diff
@@ -111,14 +111,46 @@ TEST(VariantTest, shortDecimal) {
   // 0.1234 < 1.234
   EXPECT_LT(
       u2.value<TypeKind::SHORT_DECIMAL>(), v.value<TypeKind::SHORT_DECIMAL>());
+
+  // Hash test.
```
Comment (mbasmanova) [inline velox/type/tests/VariantTest.cpp:115]: This test is getting very long. Would you extract hash testing into a separate test?

File: velox/type/tests/VariantTest.cpp [hunk 11]
```diff
@@ -141,14 +173,46 @@ TEST(VariantTest, longDecimal) {
   // 12.3456 > 12.345
   EXPECT_LT(
       v.value<TypeKind::LONG_DECIMAL>(), u2.value<TypeKind::LONG_DECIMAL>());
+
+  // Hash test.
```
Comment (mbasmanova) [inline velox/type/tests/VariantTest.cpp:177]: ditto

**Review by mbasmanova — APPROVED** [review]
@karteekmurthys Looks good to me % a few comments.

**Comment by majetideepak** [comment]
I am working on addressing the review comments.

**Comment by Yuhta** [comment]
So with this hash implementation, two nulls with different precision/scale is considered different?

**Comment by majetideepak** [comment]
@Yuhta that is right. It is safe if NULL is tied to precision, scale like a regular unscaled value.

**Review by Yuhta — APPROVED** [review]

**Pull request closed without merging**